lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -514,15 +534,41 @@ def post(self):
         are not in the class schema are ignored but logged.
         """
         json_data = request.get_json(silent=True)
+        if json_data and "user" in json_data:
+            if not json_data["user"]:
+                # There is an empty user field present in the json data;
+                # therefore, this request will be interpreted as a
+                # public dataset query.  Remove it before validating the
+                # remaining json data.
+                del json_data["user"]
+            elif Auth.token_auth.current_user() is None:
+                # Only logged-in users are allowed to query non-public data.
+                abort(HTTPStatus.FORBIDDEN, message="Not Authorized")
         try:
             new_data = self.schema.validate(json_data)
+            # After the json data is validated make sure the user parameter
+            # is present in new_data for an endpoint which needs a "user"
+            # parameter. If the request does not contain the user field,
+            # it will be interpreted as a public dataset query. Therefore
+            # after validating the json data we will manually add the user
+            # field as None to let the assemble method understand that this
+            # query is for the public dataset.
+            if (
+                "user" in self.schema.parameters.keys()
+                and "user" not in new_data.keys()
+            ):
+                new_data["user"] = None
```

Comment:
This seems odd.  Why do we have to do this when we just removed it above?

Reply:
That is because the `assemble` method of the query APIs need the "user" field explicitly set as `None` for querying a public dataset. Therefore after validating the rest of the json data we add it back as a `None`.  The reason I removed it before validating is because the user field was present but empty and validator would raise it as an exception since its not a valid user field however we interpret it as a public dataset query.

But you raised a valid point we dont need to do this, the assemble method can do `user = json_data.get("user")` instead of `user = json_data["user"]` and that should take care of setting the local `user` variable as None.